service/header/store.go

```diff
@@ -0,0 +1,318 @@
+package header
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+
+	"github.com/tendermint/tendermint/libs/bytes"
+)
+
+// TODO(@Wondertan): Those values must be configurable and proper defaults should be set for specific node type.
+var (
+	// DefaultStoreCache defines the amount of max entries allowed in the Header Store cache.
+	DefaultStoreCacheSize = 1024
+	// DefaultIndexCache defines the amount of max entries allowed in the Height to Hash index cache.
+	DefaultIndexCacheSize = 256
+)
+
+type store struct {
+	ds    datastore.Batching
+	cache *lru.ARCCache
+	index *heightIndexer
+
+	headLk sync.RWMutex
+	head   bytes.HexBytes
+}
+
+// NewStore constructs a Store over datastore.
+// The datastore must have a head there otherwise Start will error.
+// For first initialization of Store use NewStoreWithHead.
+func NewStore(ds datastore.Batching) (Store, error) {
+	return newStore(ds)
+}
+
+// NewStoreWithHead initiates a new Store and forcefully sets a given trusted header as head.
+func NewStoreWithHead(ds datastore.Batching, head *ExtendedHeader) (Store, error) {
```

Comment:
I wish we could have one foolproof constructor instead.

Reply:
I don't like passing nil into constructors and this would be the case in a unified constructor if a head exists.

Reply:
But I agree this is not good enough and there might be a better UX for this

Reply:
For example, we can have one constructor and a helper function that will inject a new Head into Datastore.